modules/tensor_mechanics/src/materials/PlasticTruss.C

```diff
@@ -0,0 +1,152 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PlasticTruss.h"
+#include "Function.h"
+#include "MooseException.h"
+#include "MathUtils.h"
+
+registerMooseObject("TensorMechanicsApp", PlasticTruss);
+
+template <>
+InputParameters
+validParams<PlasticTruss>()
+{
+  InputParameters params = validParams<LinearElasticTruss>();
+  params.addClassDescription("Computes the stress and strain in a truss element considering J2 "
+                             "plasticity with linear hardening.");
+  params.addRequiredParam<Real>(
+      "yield_stress", "Input data for yield stress after which plastic strain starts accumulating");
+  params.addParam<Real>("hardening_constant", 0.0, "Hardening slope");
+  params.addParam<FunctionName>("hardening_function",
+                                "True stress as a function of plastic strain");
+  params.addParam<Real>(
+      "absolute_tolerance", 1e-10, "Absolute convergence tolerance for Newton iteration");
+  params.addParam<Real>(
+      "relative_tolerance", 1e-8, "Relative convergence tolerance for Newton iteration");
+  return params;
+}
+
+PlasticTruss::PlasticTruss(const InputParameters & parameters)
+  : LinearElasticTruss(parameters),
+    _yield_stress(getParam<Real>("yield_stress")), // Read from input file
+    _hardening_constant(getParam<Real>("hardening_constant")),
+    _hardening_function(isParamValid("hardening_function") ? &getFunction("hardening_function")
+                                                           : NULL),
+    _hardening_slope(0.0),
+    _absolute_tolerance(parameters.get<Real>("absolute_tolerance")),
+    _relative_tolerance(parameters.get<Real>("relative_tolerance")),
+    _total_stretch_old(getMaterialPropertyOld<Real>(_base_name + "total_stretch")),
+    _plastic_strain(declareProperty<Real>(_base_name + "plastic_stretch")),
+    _plastic_strain_old(getMaterialPropertyOld<Real>(_base_name + "plastic_stretch")),
+    _elastic_strain_old(getMaterialPropertyOld<Real>(_base_name + "elastic_stretch")),
+    _stress_old(getMaterialPropertyOld<Real>(_base_name + "axial_stress")),
+    _strain_increment(declareProperty<Real>(_base_name + "strain_increment_truss")),
```

Comment:
It doesn't look like we're appending `_truss` to most of the other material properties. I think we should remove that from several of the properties here for consistency.

Reply:
It's only done for the property 'strain_increament' to avoid conflicting property names between truss and continuum elements. Otherwise, we have to specify the 'base_name' all the time.

Reply:
Are you sure it's a problem if they conflict?  I think it's not an issue because they would be on different blocks, but I could be wrong.